### PR TITLE
V02 release

### PR DESCRIPTION
**Pull request by ylacombe**

### CODE REVIEW

File: INFERENCE.md [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+    model_name,
+    attn_implementation=attn_implementation
+).to(torch_device, dtype=torch_dtype)
+```
+
+## Compilation
+
+[Compiling](https://pytorch.org/docs/stable/generated/torch.compile.html) the forward method of Parler can speed up generation time by up to 4.5x.
+
+As an indication, `mode=default` brings a speed-up of 1.4 times compared to no compilation, while a more advanced mode brings much faster generation, at the cost of a longer compilation time and the need to generate twice to see the benefits of compilation.
```
Comment (sanchit-gandhi) [inline INFERENCE.md:38]:
```suggestion
As an indication, `mode=default` brings a speed-up of 1.4 times compared to no compilation, while `mode="reduce-overhead"` brings much faster generation, at the cost of a longer compilation time and the need to generate twice to see the benefits of compilation.
```

File: INFERENCE.md [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+model_name = "parler-tts/parler-tts-mini-v1"
+
+# need to set padding max length
+max_length = 50
+
+# load model and tokenizer
+tokenizer = AutoTokenizer.from_pretrained(model_name) 
+model = ParlerTTSForConditionalGeneration.from_pretrained(
+    model_name,
+    attn_implementation="eager"
```
Comment (sanchit-gandhi) [inline INFERENCE.md:56]: IIRC `attn_implementation="sdpa"` with the C++ attn kernel was fastest? Perhaps @eustlb can confirm here!

File: INFERENCE.md [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+model.forward = torch.compile(model.forward, mode=compile_mode)
+
+# warmup
+inputs = tokenizer("This is for compilation", return_tensors="pt", padding="max_length", max_length=max_length).to(device)
+
+model_kwargs = {**inputs, prompt_input_ids: inputs.input_ids, prompt_attention_mask: inputs.attention_mask, }
+
+n_steps = 1 if compile_mode == "default" else 2
+for _ in range(n_steps):
+    _ = model.generate(**model_kwargs)
```
Comment (sanchit-gandhi) [inline INFERENCE.md:71]:
```suggestion
    with sdpa_kernel(SDPBackend.MATH):
        _ = model.generate(**model_kwargs)
```

File: INFERENCE.md [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+```
+
+## Compilation
+
+[Compiling](https://pytorch.org/docs/stable/generated/torch.compile.html) the forward method of Parler can speed up generation time by up to 4.5x.
+
+As an indication, `mode=default` brings a speed-up of 1.4 times compared to no compilation, while a more advanced mode brings much faster generation, at the cost of a longer compilation time and the need to generate twice to see the benefits of compilation.
+
+```py
+import torch
```
Comment (sanchit-gandhi) [inline INFERENCE.md:41]:
```suggestion
import torch
from torch.nn.attention import sdpa_kernel, SDPBackend
```

File: INFERENCE.md [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+
+# compile the forward pass
+compile_mode = "default" # chose "reduce-overhead" for 3 to 4x speed-up
+model.generation_config.cache_implementation = "static"
+model.forward = torch.compile(model.forward, mode=compile_mode)
+
+# warmup
+inputs = tokenizer("This is for compilation", return_tensors="pt", padding="max_length", max_length=max_length).to(device)
+
+model_kwargs = {**inputs, prompt_input_ids: inputs.input_ids, prompt_attention_mask: inputs.attention_mask, }
```
Comment (sanchit-gandhi) [inline INFERENCE.md:67]: Should this be a dict?
```suggestion
model_kwargs = {**inputs, "prompt_input_ids": inputs.input_ids, "prompt_attention_mask": inputs.attention_mask, }
```

File: INFERENCE.md [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+
+n_steps = 1 if compile_mode == "default" else 2
+for _ in range(n_steps):
+    _ = model.generate(**model_kwargs)
+
+
+# now you can benefit from compilation speed-ups
+...
```
Comment (sanchit-gandhi) [inline INFERENCE.md:75]:
```suggestion
prompt = "Hey, how are you doing today?"
description = "A female speaker with a slightly low-pitched voice delivers her words quite expressively, in a very confined sounding environment with clear audio quality. She speaks very fast."

inputs = tokenizer(prompt, return_tensors="pt", padding="max_length", max_length=max_length).to(device)
prompt_inputs = tokenizer(description, return_tensors="pt").input_ids.to(device)

inputs = tokenizer("", return_tensors="pt", padding="max_length", max_length=max_length).to(device)

model_kwargs = {**inputs, "prompt_input_ids": prompt_inputs.input_ids, "prompt_attention_mask": prompt_inputs.attention_mask}

with sdpa_kernel(SDPBackend.MATH):
    audio_arr = model.generate(**model_kwargs)
```

File: INFERENCE.md [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+For example, after 86 steps we have the first second of audio ready, and so can play this without waiting for the remaining decoding steps to be complete. As we continue to generate with the Parler-TTS model, we append new chunks of generated audio to our output waveform on-the-fly. After the full 1720 decoding steps, the generated audio is complete, and is composed of 20 chunks of audio, each corresponding to 86 tokens.
+This method of playing incremental generations reduces the latency of the Parler-TTS model from the total time to generate 1720 tokens, to the time taken to play the first chunk of audio (86 tokens). This can result in significant improvements to perceived latency,  particularly when the chunk size is chosen to be small. In practice, the chunk size should be tuned to your device: using a smaller chunk size will mean that the first chunk is ready faster, but should not be chosen so small that the model generates slower than the time it takes to play the audio.
+
+
+### How Can I Use It?
+
+We've added [ParlerTTSStreamer](https://github.com/huggingface/parler-tts/blob/main/parler_tts/streamer.py) to the library. Don't hesitate to adapt it to your use-case.
```
Comment (sanchit-gandhi) [inline INFERENCE.md:96]: Handy!

File: INFERENCE.md [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+# now you can do
+text = "This is a test of the streamer class"
+description = "Jon's talking really fast."
+
+chunk_size_in_s = 0.5
+
+for (sampling_rate, audio_chunk) in generate(text, description, chunk_size_in_s):
+  # You can do everything that you need with the chunk now
+  # For example: stream it, save it, play it.
+  print(audio_chunk.shape)
```
Comment (sanchit-gandhi) [inline INFERENCE.md:159]: Maybe we can link the streaming Gradio demo here to give users an example of how it's used in practice?

File: INFERENCE.md [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+
+model = ParlerTTSForConditionalGeneration.from_pretrained(repo_id).to("cuda")
+tokenizer = AutoTokenizer.from_pretrained(repo_id, padding_side="left")
+feature_extractor = AutoFeatureExtractor.from_pretrained(repo_id)
+
+description = "A male speaker with a monotone and high-pitched voice is delivering his speech at a really low speed in a confined environment."
+input_text = "Hey, how are you doing?"
+
+inputs = tokenizer([description, description], return_tensors="pt", padding=True).to("cuda")
+prompt = tokenizer([input_text, "I'm not sure how to feel about it."], return_tensors="pt", padding=True).to("cuda")
```
Comment (sanchit-gandhi) [inline INFERENCE.md:184]:
```suggestion
input_text = ["Hey, how are you doing?", "I'm not sure how to feel about it."]
description = 2 * ["A male speaker with a monotone and high-pitched voice is delivering his speech at a really low speed in a confined environment."]

inputs = tokenizer(description, return_tensors="pt", padding=True).to("cuda")
prompt = tokenizer(input_text, return_tensors="pt", padding=True).to("cuda")
```

File: README.md [hunk 10]
```diff
@@ -34,43 +42,81 @@ pip3 install --pre torch torchaudio --index-url https://download.pytorch.org/whl
...
 sf.write("parler_tts_out.wav", audio_arr, model.config.sampling_rate)
 ```
 
+**Tips**:
+* We've set up an [inference guide](INFERENCE.md) to make generation faster. Think SDPA, torch.compile and streaming!
+* Include the term "very clear audio" to generate the highest quality audio, and "very noisy audio" for high levels of background noise
+* Punctuation can be used to control the prosody of the generations, e.g. use commas to add small breaks in speech
+* The remaining speech features (gender, speaking rate, pitch and reverberation) can be controlled directly through the prompt
+
 https://github.com/huggingface/parler-tts/assets/52246514/251e2488-fe6e-42c1-81cd-814c5b7795b0
```
Comment (sanchit-gandhi) [inline README.md:115]: Only a suggestion: but adding this as it's own separate sub-heading might bring it more visibility, and separate out the performance improvements (e.g. prompt engineering) to the inference improvements

```suggestion
### Optimizing Inference Speed

We've set up an [inference guide](INFERENCE.md) to make generation faster. Think SDPA, torch.compile and streaming!
```

File: README.md [hunk 11]
```diff
@@ -34,43 +42,81 @@ pip3 install --pre torch torchaudio --index-url https://download.pytorch.org/whl
...
 prompt_input_ids = tokenizer(prompt, return_tensors="pt").input_ids.to(device)
 
-generation = model.generate(input_ids=input_ids, prompt_input_ids=prompt_input_ids).to(torch.float32)
+generation = model.generate(input_ids=input_ids, prompt_input_ids=prompt_input_ids)
 audio_arr = generation.cpu().numpy().squeeze()
 sf.write("parler_tts_out.wav", audio_arr, model.config.sampling_rate)
 ```
 
+**Tips**:
+* We've set up an [inference guide](INFERENCE.md) to make generation faster. Think SDPA, torch.compile and streaming!
```
Comment (sanchit-gandhi) [inline README.md:109]:
```suggestion
```

**Review by sanchit-gandhi — APPROVED** [review]
Super clear guide 👌 Really like how you've laid out the inference tips and the MWE's

File: INFERENCE.md (hunk 8 above)
Comment (sanchit-gandhi) [reply INFERENCE.md:159]: One option would be to include a dropdown that shows a PyAudio example as well:

<details>

<summary> PyAudio Example </summary>

Show an example inspired by https://stackoverflow.com/a/32161871

</details>

Comment (ylacombe) [reply INFERENCE.md:159]: Unfortunately I can't test it easily on DGX ! (missing some utils)

File: INFERENCE.md [hunk 12]
```diff
@@ -0,0 +1,202 @@
+# Inference tips
+
+Parler-TTS benefits from a number of optimizations that can make the model up to 4 times faster. Add to this the ability to stream audio as it's being generated, and you can achieve time-to-first-audio in under 500ms on a modern GPU.
```
Comment (Vaibhavs10) [inline INFERENCE.md:3]:
```suggestion
Parler-TTS benefits from a number of optimizations that can make the model up to 4x faster. Add to this the ability to stream audio as it's being generated, and you can achieve time-to-first audio in under 500ms on a modern GPU.
```

File: INFERENCE.md [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+* [Efficient Attention Implementation](#efficient-attention-implementations)
+* [Compilation](#compilation)
+* [Streaming](#streaming)
+* [Batch generation](#batch-generation)
+
+## Efficient Attention implementations
+
+Parler-TTS supports [SDPA](https://pytorch.org/docs/master/generated/torch.nn.functional.scaled_dot_product_attention.html) and [Flash Attention 2](https://github.com/Dao-AILab/flash-attention).  
+
+SDPA is used by default and speeds up generation time by up to 1.4 times compared with eager attention.
```
Comment (Vaibhavs10) [inline INFERENCE.md:15]:
```suggestion
SDPA is used by default and speeds up generation time by up to 1.4x compared with eager attention.
```

File: INFERENCE.md [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+To switch between attention implementations, simply specify `attn_implementation=attn_implementation` when loading the checkpoints:
+
+```py
+from parler_tts import ParlerTTSForConditionalGeneration
+
+torch_device = "cuda:0"
+torch_dtype = torch.bfloat16
+model_name = "parler-tts/parler-tts-mini-v1"
+
+attn_implementation = "eager" # "sdpa" or "flash_attention_2"
```
Comment (Vaibhavs10) [inline INFERENCE.md:26]: let's make `sdpa` the default since it is faster?

File: INFERENCE.md [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+Parler-TTS supports [SDPA](https://pytorch.org/docs/master/generated/torch.nn.functional.scaled_dot_product_attention.html) and [Flash Attention 2](https://github.com/Dao-AILab/flash-attention).  
+
+SDPA is used by default and speeds up generation time by up to 1.4 times compared with eager attention.
+
+To switch between attention implementations, simply specify `attn_implementation=attn_implementation` when loading the checkpoints:
+
+```py
+from parler_tts import ParlerTTSForConditionalGeneration
+
+torch_device = "cuda:0"
```
Comment (Vaibhavs10) [inline INFERENCE.md:22]:
```suggestion
torch_device = "cuda:0" # use "mps" for Mac
```

File: INFERENCE.md [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+
+Here's how to create a generator out of the streamer.
+
+```py
+import torch
+from parler_tts import ParlerTTSForConditionalGeneration, ParlerTTSStreamer
+from transformers import AutoTokenizer
+from threading import Thread
+
+torch_device = "cuda:0"
```
Comment (Vaibhavs10) [inline INFERENCE.md:106]:
```suggestion
torch_device = "cuda:0" # Use "mps" for Mac 
```

File: README.md [hunk 17]
```diff
@@ -7,15 +7,23 @@ Contrarily to other TTS models, Parler-TTS is a **fully open-source** release. A
 This repository contains the inference and training code for Parler-TTS. It is designed to accompany the [Data-Speech](https://github.com/huggingface/dataspeech) repository for dataset annotation.
 
 > [!IMPORTANT]
-> We're proud to release [Parler-TTS Mini v0.1](https://huggingface.co/parler-tts/parler_tts_mini_v0.1), our first 600M parameter model, trained on 10.5K hours of audio data.
-> In the coming weeks, we'll be working on scaling up to 50k hours of data, in preparation for the v1 model.
+> **08/08/2024:** We are proud to release two new Parler-TTS checkpoints:
+> 1. [Parler-TTS Mini](https://huggingface.co/parler-tts/parler-tts-mini-v1), an 880M parameter model.
+> 2. [Parler-TTS Large](https://huggingface.co/parler-tts/parler-tts-large-v1), a 2.3 billion-parameter model.
```
Comment (Vaibhavs10) [inline README.md:12]:
```suggestion
> 2. [Parler-TTS Large](https://huggingface.co/parler-tts/parler-tts-large-v1), a 2.3B parameter model.
```

**Review by Vaibhavs10 — APPROVED** [review]
Nice! very comprehensive, left some suggestions to harmonise text across! let's go!

File: INFERENCE.md (hunk 2 above)
Comment (eustlb) [reply INFERENCE.md:56]: I confirm that the C++ sdpa implementation makes compiled forward with `attn_implementation="sdpa"` fastest. Yet for the mini architecture `attn_implementation="sdpa"`  with compile mode `"default"` and dtype  `bfloat16` is still the fastest.

File: INFERENCE.md (hunk 14 above)
Comment (ylacombe) [reply INFERENCE.md:26]: `sdpa` is used by default, we're showing here how to use other implementations!

File: INFERENCE.md (hunk 2 above)
Comment (eustlb) [reply INFERENCE.md:56]: Correction: 
It makes the forward faster only for modes that capture CUDA graphs, i.e `"reduce-overhead"` and `"max-autotune"` but not for "default" where is actually halved generation speed.

File: INFERENCE.md (hunk 8 above)
Comment (scenaristeur) [reply INFERENCE.md:159]: @sanchit-gandhi @ylacombe could you please give an example for playing the chunk, with pyaudio or something else, please ?
